### PR TITLE
feat: Add projects to the SDK

### PR DESCRIPTION
**Pull request by wes-turner**
New:
* `determined.common.experimental.project.Project`
  * `_from_bindings` constructor
  * setters for attributes (name and description)
  * `to_json()`
  * `list_experiments()`
* Experiment:
  * `Experiment.move_to_project(workspace_name, project_name)`
  * `Experiment` now also has `project_id` and `workspace_id` attributes
* Workspace
  * `Workspace.list_projects`
  * `Workspace.get_project(project_name)`
* `det project list <workspace_name> <project_name>` now uses SDK.
## Test Plan

Create a client and log in
```python
from determined.experimental import client
client.login(<login details>)
```

Then create objects for the various resources and play with the new methods. Ex:
```python
exp = client.get_experiment(<id>)
exp.move_to_project(<workspace>, <project>)
client.get_experiment(<id>)  # check that the project has been updated on master


ws = client.list_workspaces()[<i>]
ws.list_projects()

proj = ws.get_project(<name>)
proj.list_experiments()

proj.set_description(<desc>)
ws.list_projects()  # New objects to check that set_description changed the project on master
```

Also worth playing with the refactored CLI function:
```sh
det project list <workspace name> <project name>
```
Possible flags to exercise: `--limit`, `--offset`, `--sort-by`, `--order-by`

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by azhou-determined** [comment]
should this be pointing to `python-sdk` and not `main`? my reasoning is that all the conventions this PR builds on (`hydrate`, `reload`, etc) are a part of `python-sdk`. it seems weird for us to release this before the other python SDK work.

**Comment by wes-turner** [comment]
> should this be pointing to `python-sdk` and not `main`? my reasoning is that all the conventions this PR builds on (`hydrate`, `reload`, etc) are a part of `python-sdk`. it seems weird for us to release this before the other python SDK work.

Per earlier conversations, we're going to ask for feedback about the feature branch before releasing it. Seeing here if I can enumerate our reasons in doing so:
1. what's useful in the Q3 work we've done / direct Q4 work
2. validate the structural decisions you've cited
3. check usefulness of interface (`list_*`, etc.) before general release

For 1, I think the packaging isn't super relevant to the scope we ask users to consider.

For 2, mmm, it makes sense to have examples included there in the PR. Gives the reader some context for how those structural decisions look in a simple resource.

For 3, the SDK's interface isn't yet all that consistent. Once the feature branch in merged, consistency will be better in some ways and worse in others. We'll add/deprecate/remove names in Q4. `Project` probably won't be affected by this more or less if we wait until the release of the feature branch.

Do you think that's it?

If so, I guess I'm still leaning `main`, though I don't feel particularly strongly about it. Mostly out of a desire to make small changes and keep `main` current. The feature branch is a big patch to merge, and that makes me a little uncomfortable.

FWIW, we released [workspace](https://github.com/determined-ai/determined/blob/main/harness/determined/common/experimental/workspace.py) already to address an urgent need.

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [inline harness/determined/common/experimental/project.py:113]: bindings' `to_json` is weirdly-named. Do we want to name it any different in the SDK?

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:44]: Ooh, that's an _excellent_ idea. Totally. I agree that `n_experiments` etc. is weird, and this is a way better way to get the same information (should someone actually care).

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:8]: You're totally right. I should have done that with `Workspace` when I added it. I'll submit a separate PR for that.

I'm less convinced about projects. I'd been thinking about them as a division of workspaces that don't make sense outside that context. But if you see them different, I'll add `Determined` support for them, too.

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:127]: Fair enough. I ported the CLI's "list projects" to use the SDK (and fixed a CLI bug in the meantime (`--limit` was broken)).

Also, thanks -- this should have also been a part of the original PR. It's hard to tell how useful your library is going to be if nobody tries to use it.

File: harness/determined/common/experimental/workspace.py
Comment (wes-turner) [reply harness/determined/common/experimental/workspace.py:111]: [Style guide](https://google.github.io/styleguide/pyguide.html#274-decision) says:
> Multiple for clauses or filter expressions are not permitted. Use loops instead when things get more complicated.

I agree with it here. Nested "for"s are always hard for me to read.

I chose the `itertools` route here over a for loop because the object this line creates is really natural. Honestly, I think something like this is what `read_paginated` should return in the first place.

Which, hmm... I don't think that'd be hard to write. `limit` and `offset` would need some accommodation, but doing this still would be better.

File: harness/tests/determined/common/experimental/test_project.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_project.py:41]: Maintaining a mapping between backend name and frontend would be nice, but hydration unfortunately isn't just a simple copying (`note`, here, needs to be unboxed). I guess my ideal test would ensure that during hydration:
```python
obj.frontend_name = f(bindings.backend_name)
```
for some `f`, and during `to_json`:
```python
out.backend_name = g(frontend_name)
```
for some `g` for every `backend_name` that's hydrated.

I'm not sure how exactly to write that test. This test was a "best effort."

Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_project.py:41]: > i could see a case for hydrating an attribute that doesn't belong in to_json

What do you have in mind?

**Review by wes-turner — COMMENTED** [review]
Rebased to atop `python-sdk` for two reasons:

1. Playing with the code made me realize I didn't pick the best interfaces the first time around (also with the published `Workspace`), and I don't want to make it hard to undo those bad choices.
2. Nice to use the new `Experiment` class (instead of an `ExperimentReference`)

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:43]: First took out `n_experiments` etc. and then put them back because those attributes are one of the only attributes shown in the CLI. Maybe that means somebody cares.

File: harness/determined/common/experimental/project.py
Comment (azhou-determined) [reply harness/determined/common/experimental/project.py:45]: yeah, that sounds good. notes are basically like metadata for model/checkpoint, and that's basically what we do for those, minus the copying master notes to local (which we should do).

Comment (wes-turner) [reply harness/determined/common/experimental/project.py:45]: done

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/experiment.py:90]: project_id: (Mutable, int) The ID of the project associated with the experiment.

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/experiment.py:91]: workspace_id: (Mutable, int) The ID of the workspace associated with the experiment.

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/experiment.py:284]: """Transfer the experiment to another project.

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/experiment.py:286]: This method updates the local object and the master database with the specified project and workspace details.

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/experiment.py:291]: Args:
        project_name: The name of the destination project for the experiment.
        workspace_name: The name of the workspace containing the project.
    """

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:58]: """Set this object's mutable attributes to those in the bindings object."""

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:86]: """Set the project's description locally and on the master.
Args:
        description: New description to set.
"""

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:105]: """Set the project's name locally and on the master.
        Args:
            name: New name to set.
        """

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:115]: """Set the project to archived (archived = True) locally and on the master.
        """

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:123]: """Set the project to unarchived (archived = False) locally and on the master.
        """

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:143]: """Add a note to the project.
        Because there is no functionality on the backend to add a single note, this method fetches current notes, adds a new note, and updates the master.
        Post-execution, the object's notes attribute matches the updated master's notes, 
        possibly reflecting changes to the project that have happened since the project was last hydrated from master.
        Args:
            name: Note's title.
            contents: Note's content.
        """

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:176]: """Remove a note from the project by name.
Because there is no functionality on the backend to remove a single note, this method fetches current notes, removes the note, and updates the master.
WARNING:
Post-execution, the object's notes attribute matches the updated master's notes,
possibly reflecting changes to the project that have happened since the project was last hydrated from master.

Args:
name: Title of the note to remove.

        Raises:
            ValueError: If the note with the passed name isn't found
            NotImplementedError: If multiple notes with the passed name are found
        It is possible to for a project to have multiple notes with the same name. The method's
        interface doesn't make it possible to choose which note to remove.
        """

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [inline harness/determined/common/experimental/project.py:203]: """Return this object as a dictionary.
        Attributes may be renamed to accord with the database they're stored in.
        Named `to_json` for historical reasons.
        """

**Review by tara-det-ai — COMMENTED** [review]
suggested edits added

File: harness/determined/common/experimental/experiment.py
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:90]: I feel there's some information in "project is a part of an experiment" and "experiment runs in a workspace." These requested changes remove that. Is my conception of project/workspace/experiment wrong?

File: harness/determined/common/experimental/experiment.py
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:284]: "move" is consistent with both the CLI's `det experiment move` command and the web UI. All other things being equal, it's nice to minimize changes between the two. Where does this recommendation for "transfer" instead of "move" come from?

File: harness/determined/common/experimental/experiment.py
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:286]: Existing text is [consistent with our style guide](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods). Let's chat if this is something you'd like to change.

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:58]: "From" is slightly more accurate, here, since there's a small transformation in the bindings object's attributes that happens before setting the mutable attributes. Even without that transformation, "from" feels to me like it captures execution better. Could you help me understand your thinking?

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:86]: I like it!

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:115]: That's nice, concise language.

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:143]: I like the warning text you've suggested and have updated the docstring, but the I think the textual description of what the method does is a little less clear without the enumeration. What are your thoughts here?

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:177]: I've changed this method's interface. Could you please take a look?

File: harness/determined/common/experimental/project.py
Comment (wes-turner) [reply harness/determined/common/experimental/project.py:213]: Picked something in between, since "consistency" carries more information than "reasons."

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/experiment.py:284]: let's keep "move" since it is consistent with the CLI

(i was thinking of "moving" as motion and "transferring" as moving from one place to another)

File: harness/determined/common/experimental/experiment.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/experiment.py:286]: thanks, i added a link to the google styleguide from the documentation confluence site

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/project.py:58]: Let's keep "from" if it is more accurate. It sounded awkward when I read it out loud.

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/project.py:133]: I was going for brevity. The enumeration could help. I like to save the imperative mood for step-by-step instructions. I was thinking if the docstring gets rendered in a page of content on the docs site via Sphinx autodocs, they may, at first glance, think these are instructions. But the enumeration (or bullet points) you had is definitely clear in the context of just looking at the script.

Comment (tara-det-ai) [reply harness/determined/common/experimental/project.py:133]: I see that the google style guide also talks about imperative mood. To clarify, I use numbered steps when providing step by step instructions written in the imperative mood.

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/project.py:166]: i'll come back to this one

File: harness/determined/cli/workspace.py
Comment (azhou-determined) [inline harness/determined/cli/workspace.py:121]: can we put these defaults in `args_description`?

File: harness/determined/cli/workspace.py
Comment (azhou-determined) [inline harness/determined/cli/workspace.py:126]: what happens if `sort_key` is invalid/doesn't exist on project? should we define a default or throw up?

File: harness/determined/cli/workspace.py
Comment (azhou-determined) [inline harness/determined/cli/workspace.py:129]: hmm... this seems weird to me. i sort of doubt that offset/limit is even needed here, but if it is, should it go in list_projects/REST API?

File: harness/determined/common/experimental/workspace.py
Comment (azhou-determined) [reply harness/determined/common/experimental/workspace.py:111]: i agree nested for loops are unideal. but this basically splits up the nesting into two steps, which could also be done with list comprehension?
`bindings_projects = [r.projects for r in api.read_paginated(get_with_offset)]`?

i guess i tend to prefer minimal imports/dependencies (though itertools seems harmless). also `itertools` seems less readable than list comprehension. i have no idea if that does something special without looking it up.

either way, not a big deal.

File: harness/tests/determined/common/experimental/test_project.py
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_project.py:41]: > What do you have in mind?

nothing in particular, i guess i feel like `to_json`'s purpose is to serve the CLI (ultimately to render to users), whereas attributes of the resource object in the SDK. like what if at some point we need to introduce a private attribute on `Project` for whatever reason? it shouldn't automatically go in the json right?

actually, what if we just move `to_json` to the cli? it's feeling weird to me that `to_json` serves only one purpose in the SDK, to be rendered by the CLI.

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/project.py:133]: looks good, i'll set this to resolved

File: harness/determined/common/experimental/project.py
Comment (tara-det-ai) [reply harness/determined/common/experimental/project.py:166]: looks good. setting to resolved.

**Review by tara-det-ai — APPROVED** [review]
LGTM

File: harness/determined/cli/workspace.py
Comment (wes-turner) [reply harness/determined/cli/workspace.py:121]: A good idea that someone already beat you to: https://github.com/determined-ai/determined/blob/main/harness/determined/cli/workspace.py#L378-L391

File: harness/determined/cli/workspace.py
Comment (wes-turner) [reply harness/determined/cli/workspace.py:126]: * If `sort_key` is invalid:
  - `sort_key` must be one of `["id", "name"]` as enforced by the argument parser. I'm not categorically opposed to adding validation to the method, too, but if after hearing this you still want it then I'd like to talk about it more
* If `sort_key` doesn't exist on project
  - Could you please clarify this for me? I didn't catch your meaning.

File: harness/determined/cli/workspace.py
Comment (wes-turner) [reply harness/determined/cli/workspace.py:129]: I believe that usage in both CLI and SDK of `offset` is going to be incredibly rare, and that `limit` only somewhat more common.

Keeping methods simple until product or a customer asks for more is one of the parts of the manifesto, and I'd prefer to not implement it there. It's so easy for a user to do, themselves, if they really need it (minus the extra data/time over the wire from filtering client side).

I wouldn't have implemented it here, either, except it was already done and it just didn't feel like the time to remove functionality. Also, I don't mind as much having the flags in the CLI as in the SDK -- it's _marginally_ harder for a user to add themselves, and if you just want to glance at a couple lines it's got _some_ functionality.

I mean, I still wouldn't have offered it if it didn't already exist. I just don't have the same idea of purity as if it was SDK.

File: harness/determined/cli/workspace.py
Comment (wes-turner) [reply harness/determined/cli/workspace.py:121]: Looks like
```
% det -m latest-main.determined.ai:8080 w list-projects --help
usage: det workspace list-projects [-h] [--sort-by {id,name}] [--order-by {asc,desc}] [--limit LIMIT] [--offset OFFSET] [--json] workspace_name

positional arguments:
  workspace_name        name of the workspace

optional arguments:
  -h, --help            show this help message and exit
  --sort-by {id,name}   sort workspaces by the given field (default: id)
  --order-by {asc,desc}
                        order workspaces in either ascending or descending order (default: asc)
  --limit LIMIT         Maximum items per page of results (default: 200)
  --offset OFFSET       Number of items to skip before starting page of results (default: None)
  --json                print as JSON (default: False)
```

File: harness/tests/determined/common/experimental/test_project.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_project.py:41]: You know... you're right. I'd seen `to_json` as necessary to support the SDK, but as core functionality it's just as weird as you say.

So I went looking for where to put it into the CLI and I do not like what I see. Off-hand, it makes the most sense in `determined.cli.project`, right? But... `list-projects` is also accessed through `determined.cli.workspace`, and it feels gross to me for the various modules within `determined.cli` to import each other.

I think there's probably a better design that separates rendering and execution, but that design needs a _big_ refactor.

Until then, I'd like to just make `to_json` private and add a TODO. That cool? I'll even file a ticket if you think that's best practices.

File: harness/determined/common/experimental/workspace.py
Comment (wes-turner) [reply harness/determined/common/experimental/workspace.py:111]: If you like, sure. Done.

Comment (wes-turner) [reply harness/determined/common/experimental/workspace.py:111]: Well... undone. The type of `[r.projects for r in api.read_paginated(get_with_offset)]` is a `[[bindings.v1Project]]`, not a `[bindings.v1Project]` (list of lists, not a list). Flattening the nesting is one of the nice things `chain.from_iterable` does.

If you're less a fan of `itertools.chain.from_iterable`, then I can replace it with an accumulator:
```python
projects  = []
for binding_projects in api.read_paginated(get_with_offset):
  for p in binding_projects:
    projects.append(project.Project._from_bindings(p, self._session)
```

File: harness/tests/determined/common/experimental/test_project.py
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_project.py]: ugh, yeah, you're right. making it private is a good idea for now.

File: harness/determined/common/experimental/workspace.py
Comment (azhou-determined) [reply harness/determined/common/experimental/workspace.py:111]: nah, makes sense.

File: harness/determined/cli/workspace.py
Comment (azhou-determined) [reply harness/determined/cli/workspace.py:121]: so we don't need these defaults anymore right? i meant the defaults should be set in `args_description`, not just the help text

File: harness/determined/cli/workspace.py
Comment (azhou-determined) [reply harness/determined/cli/workspace.py:126]: i missed that it was enforced by the argument parser. cool.

File: harness/determined/cli/workspace.py
Comment (wes-turner) [reply harness/determined/cli/workspace.py:121]: ohh. I'm here only days later understanding what you meant. _These_ defaults meant the defaults that I've added inline. Since they were already there before this PR, I didn't even consider what I did when refactoring it.

Yes. Agreed. Thanks. Getting rid of them.

Comment (wes-turner) [reply harness/determined/cli/workspace.py]: Also didn't need to lowercase the sort args, since that's also enforced by the argparser.

**Review by azhou-determined — APPROVED** [review]
nice!

**Pull request merged**
